catalog/permissions/permissions.go

```diff
@@ -0,0 +1,152 @@
+	client  *databricks.WorkspaceClient
+	context context.Context
+}
+
+func NewUnityCatalogPermissionsAPI(ctx context.Context, m any) UnityCatalogPermissionsAPI {
+	client, _ := m.(*common.DatabricksClient).WorkspaceClient()
+	return UnityCatalogPermissionsAPI{client, ctx}
+}
+
+func (a UnityCatalogPermissionsAPI) GetPermissions(securable catalog.SecurableType, name string) (list *catalog.PermissionsList, err error) {
+	if securable.String() == "share" {
+		list, err = a.client.Shares.SharePermissions(a.context, sharing.SharePermissionsRequest{name})
+		return
+	}
+	list, err = a.client.Grants.GetBySecurableTypeAndFullName(a.context, securable, name)
+	return
+}
+
+func (a UnityCatalogPermissionsAPI) UpdatePermissions(securable catalog.SecurableType, name string, diff []catalog.PermissionsChange) error {
+	if securable.String() == "share" {
+		return a.client.Shares.UpdatePermissions(a.context, sharing.UpdateSharePermissions{
+			Changes: diff,
+			Name:    name,
+		})
+	}
+	_, err := a.client.Grants.Update(a.context, catalog.UpdatePermissions{
+		Changes:       diff,
+		SecurableType: securable,
+		FullName:      name,
+	})
+	return err
+}
+
+func (a UnityCatalogPermissionsAPI) WaitForUpdate(timeout time.Duration, securable catalog.SecurableType, name string, desired catalog.PermissionsList, diff func(*catalog.PermissionsList, catalog.PermissionsList) []catalog.PermissionsChange) error {
+	return retry.RetryContext(a.context, timeout, func() *retry.RetryError {
+		current, err := a.GetPermissions(securable, name)
+		if err != nil {
+			return retry.NonRetryableError(err)
+		}
+		log.Printf("[DEBUG] Permissions for %s-%s are: %v", securable.String(), name, current)
+		if diff(current, desired) == nil {
+			return nil
+		}
+		return retry.RetryableError(
+			fmt.Errorf("permissions for %s-%s are %v, but have to be %v", securable.String(), name, current, desired),
+		)
+	})
+}
+
+// Terraform Schema
+type UnityCatalogPrivilegeAssignment struct {
+	Principal  string   `json:"principal"`
+	Privileges []string `json:"privileges" tf:"slice_set"`
+}
+
+// Permission Mappings
+
+type SecurableMapping map[string]catalog.SecurableType
+
+// reuse ResourceDiff and ResourceData
+type attributeGetter interface {
+	Get(key string) any
+}
+
+func (sm SecurableMapping) GetSecurableType(securable string) catalog.SecurableType {
+	return sm[securable]
+}
+
+func (sm SecurableMapping) KeyValue(d attributeGetter) (string, string) {
+	for field := range sm {
+		v := d.Get(field).(string)
+		if v == "" {
+			continue
+		}
+		return field, v
+	}
+	return "unknown", "unknown"
+}
+func (sm SecurableMapping) Id(d *schema.ResourceData) string {
+	securable, name := sm.KeyValue(d)
+	return fmt.Sprintf("%s/%s", securable, name)
+}
+
+// Mappings
+// See https://docs.databricks.com/api/workspace/grants/update for full list
+// Omitting provider as a reserved keyword
+var Mappings = SecurableMapping{
+	"catalog":            catalog.SecurableType("catalog"),
```

Comment:
Not for this PR, but a general remark: seems like it would be useful for the SDK itself to expose a function that returns all enum values for an enum type.